src/goto-symex/complexity_module.cpp

```diff
@@ -0,0 +1,205 @@
+  // no matter the branching, that we cancel out early. This is critical because
+  // otherwise we will continue to execute nested loops that we already know
+  // are far too complicated to deal with, wasting huge amounts of time.
+  std::size_t sum_complexity = 0;
+
+  // If non-empty, is the lowest-level loop we should blacklist in the stack.
+  optionalt<std::reference_wrapper<lexical_loopst::loopt>> loop_to_blacklist;
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    for(auto it = frame_iter->active_loops.rbegin();
+        it != frame_iter->active_loops.rend();
+        it++)
+    {
+      auto &loop_info = *it;
+
+      // If we've already violated a complexity limit, add to all parent loops
+      // our currently blacklisted ones
+      if(loop_to_blacklist)
+      {
+        loop_info.blacklisted_loops.emplace_back(*loop_to_blacklist);
+      }
+      else
+      {
+        sum_complexity += loop_info.children_too_complex;
+        if(sum_complexity > max_loops_complexity)
+        {
+          loop_to_blacklist =
+            std::reference_wrapper<lexical_loopst::loopt>(loop_info.loop);
+        }
+      }
+    }
+  }
+
+  return !loop_to_blacklist;
+}
+
+complexity_violationt
+complexity_modulet::check_complexity(goto_symex_statet &state)
+{
+  if(!complexity_limits_active() || !state.reachable)
+    return complexity_violationt::NONE;
+
+  std::size_t complexity = state.complexity();
+  if(complexity == 0)
+    return complexity_violationt::NONE;
+
+  auto &current_call_stack = state.call_stack();
+
+  // Check if this branch is too complicated to continue.
+  auto active_loop = get_current_active_loop(current_call_stack);
+  if(complexity >= max_complexity)
+  {
+    // If we're too complex, add a counter to the current loop we're in and
+    // check if we've violated children complexity limits.
+    if(active_loop)
+    {
+      (*active_loop).get().children_too_complex++;
+
+      // If we're considered too complex, cancel branch.
+      if(are_loop_children_too_complicated(current_call_stack))
+      {
+        log.warning()
+          << "[symex-complexity] Loop operations considered too complex"
+          << (state.source.pc->source_location.is_not_nil()
+                ? " at: " + state.source.pc->source_location.as_string()
+                : ".")
+          << messaget::eom;
+
+        return complexity_violationt::LOOP;
+      }
+    }
+
+    log.warning() << "[symex-complexity] Branch considered too complex"
+                  << (state.source.pc->source_location.is_not_nil()
+                        ? " at: " + state.source.pc->source_location.as_string()
+                        : ".")
+                  << messaget::eom;
+
+    // Then kill this branch.
+    return complexity_violationt::BRANCH;
+  }
+
+  // If we haven't found a bottom-most active loop, just cancel.
```

Comment:
"just cancel" suggests "cancelling," i.e. truncating a symex path, as used above. In fact here we're *not* cancelling.